File: fetchPrices.js
/**
 * Module d'extraction de prix pour Expo / React Native (iOS & Android)
 * Fonctionne à 100% dans Expo Go sans backend et sans dépendance externe.
 *
 * Utilise l'API native fetch() pour récupérer les prix de :
 * Carrefour, Auchan, E.Leclerc, Intermarché, Lidl, Monoprix, etc.
 */

const OPEN_PRICES_PRODUCTS_URL = "https://prices.openfoodfacts.org/api/v1/products";
const OPEN_PRICES_ITEMS_URL = "https://prices.openfoodfacts.org/api/v1/prices";
const OFF_PRODUCT_URL = "https://world.openfoodfacts.org/api/v2/product";

const HEADERS = {
  "User-Agent": "PrixMagasin-ExpoApp/1.0 (contact: mobile@prixmagasin.local)",
};

/**
 * Normalise le nom de l'enseigne
 */
function normalizeRetailer(rawName) {
  if (!rawName) return "Autre";
  const brands = [
    "Carrefour",
    "Auchan",
    "E.Leclerc",
    "Leclerc",
    "Intermarché",
    "Lidl",
    "Monoprix",
    "Super U",
    "Système U",
    "Casino",
    "Aldi",
  ];
  for (const b of brands) {
    if (rawName.toLowerCase().includes(b.toLowerCase())) {
      return b.toLowerCase().includes("leclerc") ? "E.Leclerc" : b;
    }
  }
  return "Autre";
}

/**
 * Récupère les prix et détails d'un produit.
 * @param {string} queryOrEan - Soit un mot-clé (ex: "Nutella"), soit un code-barres (ex: "3017620422003")
 * @param {number} maxResults - Nombre max de produits à renvoyer (défaut: 5)
 * @returns {Promise<Array>} Liste des produits avec leurs prix par supermarché
 */
export async function getProductPrices(queryOrEan, maxResults = 5) {
  try {
    const query = (queryOrEan || "").trim();
    if (!query) return [];

    const isEan = /^\d{8,14}$/.test(query);
    const productsToCheck = [];

    if (isEan) {
      // 1. Recherche directe par code-barres (idéal avec le scanner Expo Camera)
      const offRes = await fetch(`${OFF_PRODUCT_URL}/${query}.json`, { headers: HEADERS });
      const offData = await offRes.json();
      const p = offData.product || {};
      productsToCheck.push({
        ean: query,
        name: p.product_name_fr || p.product_name || "Produit inconnu",
        brand: p.brands || "Marque inconnue",
        imageUrl: p.image_front_small_url || p.image_url || null,
        nutriscore: p.nutriscore_grade || null,
      });
    } else {
      // 2. Recherche par nom dans le catalogue de prix
      const url = `${OPEN_PRICES_PRODUCTS_URL}?product_name__like=${encodeURIComponent(query)}&size=${maxResults}&order_by=-price_count`;
      const res = await fetch(url, { headers: HEADERS });
      const data = await res.json();
      for (const item of data.items || []) {
        productsToCheck.push({
          ean: item.code,
          name: item.product_name,
          brand: item.brands || "Marque",
          imageUrl: item.image_url || null,
          nutriscore: item.nutriscore_grade && item.nutriscore_grade !== "unknown" ? item.nutriscore_grade : null,
        });
      }
    }

    // 3. Pour chaque produit, récupérer les prix réels constatés par supermarché
    const finalResults = [];

    for (const prod of productsToCheck) {
      if (!prod.ean) continue;

      const pricesUrl = `${OPEN_PRICES_ITEMS_URL}?product_code=${prod.ean}&size=25&order_by=-date`;
      const pRes = await fetch(pricesUrl, { headers: HEADERS });
      const pData = await pRes.json();

      const pricesByRetailer = {};
      for (const item of pData.items || []) {
        const loc = item.location || {};
        const rawName = loc.osm_name || loc.osm_brand || "Magasin";
        const retailer = normalizeRetailer(rawName);
        const priceVal = item.price;

        if (priceVal !== null && priceVal !== undefined) {
          const numPrice = Number(parseFloat(priceVal).toFixed(2));
          // Garder le prix le plus bas pour chaque enseigne
          if (!pricesByRetailer[retailer] || numPrice < pricesByRetailer[retailer].price) {
            pricesByRetailer[retailer] = {
              retailer: retailer,
              price: numPrice,
              storeName: loc.osm_name || retailer,
              city: loc.osm_address_city || "France",
              date: item.date || null,
              isDiscounted: Boolean(item.price_is_discounted),
            };
          }
        }
      }

      // Trier du moins cher au plus cher
      const sortedPrices = Object.values(pricesByRetailer).sort((a, b) => a.price - b.price);
      const bestPrice = sortedPrices.length > 0 ? sortedPrices[0].price : null;
      const bestRetailer = sortedPrices.length > 0 ? sortedPrices[0].retailer : null;

      finalResults.push({
        ean: prod.ean,
        name: prod.name,
        brand: prod.brand,
        imageUrl: prod.imageUrl,
        nutriscore: prod.nutriscore,
        bestPrice: bestPrice,
        bestRetailer: bestRetailer,
        prices: sortedPrices,
      });
    }

    return finalResults;
  } catch (error) {
    console.error("Erreur getProductPrices:", error);
    return [];
  }
}

File: fetch_prices.py
"""
Module autonome d'extraction et comparaison de prix multi-supermarchés en France.
(Auchan, Carrefour, E.Leclerc, Intermarché, Lidl, Monoprix...)

Usage en ligne de commande :
    python fetch_prices.py "Nutella"
    python fetch_prices.py "3017620422003"

Usage en import Python dans une autre application :
    from fetch_prices import get_product_prices
    
    resultats = get_product_prices("Nutella")
    print(resultats)
"""

import urllib.request
import urllib.parse
import json
import re
import sys

# Compatibilité console Windows (UTF-8)
if sys.stdout and sys.stdout.encoding and sys.stdout.encoding.lower() != 'utf-8':
    try:
        sys.stdout.reconfigure(encoding='utf-8')
    except Exception:
        pass


OPEN_PRICES_SEARCH_URL = "https://prices.openfoodfacts.org/api/v1/products"
OPEN_PRICES_ITEMS_URL = "https://prices.openfoodfacts.org/api/v1/prices"
OFF_PRODUCT_URL = "https://world.openfoodfacts.org/api/v2/product/{ean}.json"

HEADERS = {
    "User-Agent": "PrixMagasin-Client/1.0 (contact: support@prixmagasin.local)"
}

def _http_get_json(url: str, params: dict = None) -> dict:
    """Effectue une requête GET HTTP et renvoie le JSON parsé."""
    full_url = url
    if params:
        full_url = f"{url}?{urllib.parse.urlencode(params)}"
    req = urllib.request.Request(full_url, headers=HEADERS)
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            return json.loads(resp.read().decode('utf-8'))
    except Exception as e:
        return {}

def get_product_prices(query_or_ean: str, max_results: int = 5) -> list:
    """
    Recherche un produit par mot-clé (ex: 'Nutella') ou code EAN (ex: '3017620422003')
    et renvoie la liste des produits avec leurs prix comparés par enseigne.
    """
    query = query_or_ean.strip()
    is_ean = query.isdigit() and len(query) in (8, 12, 13, 14)

    products_to_check = []

    if is_ean:
        # Recherche directe par code-barres
        off_data = _http_get_json(OFF_PRODUCT_URL.format(ean=query))
        prod = off_data.get("product", {})
        products_to_check.append({
            "ean": query,
            "name": prod.get("product_name_fr") or prod.get("product_name") or "Produit",
            "brand": prod.get("brands", "Marque inconnue"),
            "image_url": prod.get("image_front_small_url") or prod.get("image_url"),
            "nutriscore": prod.get("nutriscore_grade")
        })
    else:
        # Recherche textuelle dans le catalogue de produits
        res = _http_get_json(OPEN_PRICES_SEARCH_URL, {
            "product_name__like": query,
            "size": max_results,
            "order_by": "-price_count"
        })
        for item in res.get("items", []):
            products_to_check.append({
                "ean": item.get("code"),
                "name": item.get("product_name"),
                "brand": item.get("brands", "Marque inconnue"),
                "image_url": item.get("image_url"),
                "nutriscore": item.get("nutriscore_grade")
            })

    # Pour chaque produit, récupérer les prix relevés dans les supermarchés
    final_results = []
    for prod in products_to_check:
        ean = prod["ean"]
        if not ean:
            continue

        prices_data = _http_get_json(OPEN_PRICES_ITEMS_URL, {
            "product_code": ean,
            "size": 20,
            "order_by": "-date"
        })

        prices_by_retailer = {}
        for p in prices_data.get("items", []):
            loc = p.get("location") or {}
            raw_name = loc.get("osm_name") or loc.get("osm_brand") or "Autre"

            # Normaliser le nom de l'enseigne
            retailer = "Autre"
            for brand in ["Carrefour", "Auchan", "E.Leclerc", "Leclerc", "Intermarché", "Lidl", "Monoprix", "Super U", "Casino", "Aldi"]:
                if brand.lower() in raw_name.lower():
                    retailer = "E.Leclerc" if "leclerc" in brand.lower() else brand
                    break

            price = p.get("price")
            if price is not None:
                price_val = round(float(price), 2)
                # Garder le prix le plus bas ou le plus récent pour cette enseigne
                if retailer not in prices_by_retailer or price_val < prices_by_retailer[retailer]["price"]:
                    prices_by_retailer[retailer] = {
                        "retailer": retailer,
                        "price": price_val,
                        "store_name": loc.get("osm_name"),
                        "city": loc.get("osm_address_city") or "France",
                        "date": p.get("date"),
                        "is_discounted": bool(p.get("price_is_discounted"))
                    }

        sorted_prices = sorted(list(prices_by_retailer.values()), key=lambda x: x["price"])
        best_price = sorted_prices[0]["price"] if sorted_prices else None
        best_retailer = sorted_prices[0]["retailer"] if sorted_prices else None

        final_results.append({
            "ean": ean,
            "name": prod["name"],
            "brand": prod["brand"],
            "image_url": prod["image_url"],
            "nutriscore": prod["nutriscore"],
            "best_price": best_price,
            "best_retailer": best_retailer,
            "prices": sorted_prices
        })

    return final_results

if __name__ == "__main__":
    # Test CLI
    search_term = sys.argv[1] if len(sys.argv) > 1 else "Nutella"
    print(f"--- Recherche des prix pour : '{search_term}' ---")
    data = get_product_prices(search_term, max_results=3)

    for item in data:
        print(f"\n📦 {item['name']} ({item['brand']}) [EAN: {item['ean']}]")
        print(f"   🏆 Moins cher chez : {item['best_retailer']} à {item['best_price']} €")
        print("   Prix constatés par supermarché :")
        for pr in item["prices"]:
            print(f"     • {pr['retailer']}: {pr['price']} € ({pr['city']} - {pr['date']})")
